.github/workflows/build.yml

```diff
@@ -146,58 +101,168 @@ jobs:
+            vpr/libvpr.a
+            vpr/vpr
+            yosys/share/
+            yosys/yosys
+            yosys/yosys-abc
+            yosys/yosys-config
+            yosys/yosys-filterlib
+            yosys/yosys-smtbmc
+  docker_distribution:
+    name: Build docker image for distribution
+    if: true #${{ github.ref == 'refs/heads/master' }}
+    runs-on: ubuntu-latest
+    needs: linux_build
+    steps:
+      - name: Checkout OpenFPGA repo
+        uses: actions/checkout@v2
+      - name: Download a built artifacts
+        uses: actions/download-artifact@v2
+        with:
+          name: openfpga
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      - name: Login to GitHub Container Registry
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.CR_PAT }}
+      - name: Build and push master image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./docker/Dockerfile.master
+          push: true
+          tags: ghcr.io/lnis-uofu/openfpga-master:latest
+  artifact_regression_tests:
+    name: Regression tests on code changes
+    runs-on: ubuntu-18.04
+    container: ghcr.io/lnis-uofu/openfpga-env
+    needs: linux_build
+    strategy:
+      fail-fast: false
+      matrix:
+        config:
+          - name: "Basic Regression Tests: GCC-8 (Ubuntu 18.04)"
+            script: basic_reg_test.sh
+          - name: "FPGA-Verilog Regression Tests: GCC-8 (Ubuntu 18.04)"
+            script: fpga_verilog_reg_test.sh
+          - name: "FPGA-Bitstream Regression Tests: GCC-8 (Ubuntu 18.04)"
+            script: fpga_bitstream_reg_test.sh
+          - name: "FPGA-SDC Regression Tests: GCC-8 (Ubuntu 18.04)"
+            script: fpga_sdc_reg_test.sh
+          - name: "FPGA-SPICE Regression Tests: GCC-8 (Ubuntu 18.04)"
+            script: fpga_spice_reg_test.sh
+    steps:
+      - name: Checkout OpenFPGA repo
+        uses: actions/checkout@v2
+      - name: Download a built artifacts
+        uses: actions/download-artifact@v2
+        with:
+          name: openfpga
+      - name: chmod
         run: |
-          execute_process(COMMAND ccache -s)
-
+          chmod +x abc/abc
+          chmod +x ace2/ace
+          chmod +x openfpga/openfpga
+          chmod +x vpr/vpr
+          chmod +x yosys/yosys
+          chmod +x yosys/yosys-abc
+          chmod +x yosys/yosys-config
+          chmod +x yosys/yosys-filterlib
+          chmod +x yosys/yosys-smtbmc
       - name: ${{matrix.config.name}}
-        if: contains(matrix.config.name, 'Regression Test')
         shell: bash
-        # Execute the test. 
-        run: source ./.github/workflows/${{matrix.config.reg_script}}
+        run: source .github/workflows/${{matrix.config.script}}
+  docker_regression_tests:
```

Comment:
you can replace this strategy to this,
This is how I intend to use the docker image.
```
docker_regression_tests:
    needs: change_detect
    if: ${{ !fromJSON(needs.change_detect.outputs.source_modified) }}
    name: Regression tests against master artifacts
    runs-on: ubuntu-18.04
    container: ghcr.io/lnis-uofu/openfpga-master:latest
    strategy:
      fail-fast: false
      matrix:
        config:
          - name: "Basic Regression Tests: GCC-8 (Ubuntu 18.04)"
            script: basic_reg_test.sh
          - name: "FPGA-Verilog Regression Tests: GCC-8 (Ubuntu 18.04)"
            script: fpga_verilog_reg_test.sh
          - name: "FPGA-Bitstream Regression Tests: GCC-8 (Ubuntu 18.04)"
            script: fpga_bitstream_reg_test.sh
          - name: "FPGA-SDC Regression Tests: GCC-8 (Ubuntu 18.04)"
            script: fpga_sdc_reg_test.sh
          - name: "FPGA-SPICE Regression Tests: GCC-8 (Ubuntu 18.04)"
            script: fpga_spice_reg_test.sh
    steps:
      - name: Checkout OpenFPGA repo
        uses: actions/checkout@v2
        with:
          path: ../OpenFPGA-Repo
      - name: ${{matrix.config.name}}
        shell: bash
        run: |
          cp -r ../OpenFPGA-Repo/.github .
          cp -r ../OpenFPGA-Repo/openfpga_flow .
          source .github/workflows/${{matrix.config.script}}
```